### PR TITLE
[Merged by Bors] - feat: port CategoryTheory.Closed.Cartesian

### PR DESCRIPTION
**Pull request by Vtec234**
---

- [x] depends on: #4826

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by semorrison** [comment]
This PR/issue depends on:
* leanprover-community/mathlib4#4826
By **[Dependent Issues](https://github.com/z0al/dependent-issues)** (🤖). Happy coding!

**Comment by semorrison** [comment]
As far as I can see this doesn't actually depend on https://github.com/leanprover-community/mathlib4/pull/4826!

**Comment by semorrison** [comment]
@Vtec234, please merge the dependent PR in, so CI can run.

**Comment by semorrison** [comment]
This PR/issue depends on:
* ~~leanprover-community/mathlib4#4826~~
By **[Dependent Issues](https://github.com/z0al/dependent-issues)** (🤖). Happy coding!

File: Mathlib/CategoryTheory/Closed/Cartesian.lean [hunk 1]
```diff
@@ -105,9 +105,28 @@ abbrev coev : 𝟭 C ⟶ prod.functor.obj A ⋙ exp A :=
 #align category_theory.exp.coev CategoryTheory.exp.coev
 
 -- mathport name: «expr ⟹ »
+-- porting note: notation fails to elaborate with `quotPrecheck` on.
+set_option quotPrecheck false in
 notation:20 A " ⟹ " B:19 => (exp A).obj B
 
+open Lean PrettyPrinter.Delaborator SubExpr in
+@[delab app.Prefunctor.obj]
+def delabPrefunctorObjExp : Delab := do
```
Comment (Vtec234) [inline Mathlib/CategoryTheory/Closed/Cartesian.lean:114]: I added this to display the internal hom notation when fully applied.

File: Mathlib/CategoryTheory/Closed/Cartesian.lean [hunk 2]
```diff
@@ -303,29 +330,28 @@ def mulZero {I : C} (t : IsInitial I) : I ⨯ A ≅ I :=
...
-  inv := CartesianClosed.curry ((mulZero t).Hom ≫ t.to _)
-  hom_inv_id' := by
-    rw [← curry_natural_left, curry_eq_iff, ← cancel_epi (MulZeroClass.mul_zero t).inv]
-    · apply t.hom_ext
-    · infer_instance
-    · infer_instance
+  hom := default
+  inv := CartesianClosed.curry ((mulZero t).hom ≫ t.to _)
+  hom_inv_id := by
+    -- Porting note: mathport thought that the `mulZero` here was `MulZeroClass.mul_zero`!
```
Comment (Vtec234) [inline Mathlib/CategoryTheory/Closed/Cartesian.lean:336]: This was a weird mistranslation.

File: Mathlib/CategoryTheory/Closed/Cartesian.lean [hunk 3]
```diff
@@ -0,0 +1,433 @@
...
+abbrev ev : exp A ⋙ prod.functor.obj A ⟶ 𝟭 C :=
+  ihom.ev A
+#align category_theory.exp.ev CategoryTheory.exp.ev
+
+/-- The coevaluation natural transformation. -/
+abbrev coev : 𝟭 C ⟶ prod.functor.obj A ⋙ exp A :=
+  ihom.coev A
+#align category_theory.exp.coev CategoryTheory.exp.coev
+
+-- mathport name: «expr ⟹ »
```
Comment (semorrison) [inline Mathlib/CategoryTheory/Closed/Cartesian.lean:107]:
```suggestion
```

File: Mathlib/CategoryTheory/Closed/Cartesian.lean [hunk 4]
```diff
@@ -0,0 +1,433 @@
...
+    let e ← getExpr
+    guard $ e.isAppOfArity' ``Functor.toPrefunctor 5
+    withNaryArg 4 do
+      let e ← getExpr
+      guard $ e.isAppOfArity' ``exp 5
+      withNaryArg 2 delab
+  let B ← withNaryArg 5 delab
+  `($A ⟹ $B)
+
+-- mathport name: «expr ^^ »
```
Comment (semorrison) [inline Mathlib/CategoryTheory/Closed/Cartesian.lean:127]:
```suggestion
```

File: Mathlib/CategoryTheory/Closed/Cartesian.lean [hunk 5]
```diff
@@ -0,0 +1,433 @@
...
+
+/-- Transport the property of being cartesian closed across an equivalence of categories.
+
+Note we didn't require any coherence between the choice of finite products here, since we transport
+along the `prodComparison` isomorphism.
+-/
+def cartesianClosedOfEquiv (e : C ≌ D) [h : CartesianClosed C] : CartesianClosed D
+    where closed X :=
+    {
+      isAdj := by
```
Comment (semorrison) [inline Mathlib/CategoryTheory/Closed/Cartesian.lean:405]:
```suggestion
def cartesianClosedOfEquiv (e : C ≌ D) [h : CartesianClosed C] : CartesianClosed D where 
  closed X :=
    { isAdj := by
```
and presumably then reduce indenting below.

**Comment by semorrison** [comment]
bors d+

**Pull request closed without merging**